server/repository/src/db_diesel/temperature_excursion.rs

```diff
@@ -0,0 +1,313 @@
+use chrono::{NaiveDateTime, Utc};
+
+use diesel::{prelude::*, sql_types::Integer};
+
+use super::temperature_excursion_row::TemperatureExcursionRow;
+
+pub struct TemperatureExcursionRepository<'a> {
+    connection: &'a StorageConnection,
+}
+
+type QueryResult = (
+    NaiveDateTime,
+    f64,
+    String,
+    String,
+    Option<String>,
+    i32,
+    bool,
+);
+
+#[derive(Debug, PartialEq, Clone)]
+pub struct TemperatureRow {
+    pub datetime: NaiveDateTime,
+    pub temperature: f64,
+    pub store_id: String,
+    pub sensor_id: String,
+    pub location_id: Option<String>,
+    pub is_excursion: bool,
+    pub duration: i64,
+}
+
+impl<'a> TemperatureExcursionRepository<'a> {
+    pub fn new(connection: &'a StorageConnection) -> Self {
+        TemperatureExcursionRepository { connection }
+    }
+
+    /// Result is sorted by datetime descending
+    pub fn query(
+        &self,
+        filter: TemperatureLogFilter,
+    ) -> Result<Vec<TemperatureExcursionRow>, RepositoryError> {
+        let mut query = temperature_log_dsl::temperature_log
+            .inner_join(
+                temperature_breach_config_dsl::temperature_breach_config
+                    .on(temperature_log_dsl::store_id.eq(temperature_breach_config_dsl::store_id)),
+            )
+            .select((
+                temperature_log_dsl::datetime,
+                temperature_log_dsl::temperature,
+                temperature_log_dsl::store_id,
+                temperature_log_dsl::sensor_id,
+                temperature_log_dsl::location_id,
+                (temperature_breach_config_dsl::duration_milliseconds / 1000).into_sql::<Integer>(),
+                temperature_log_dsl::temperature.not_between(
+                    temperature_breach_config_dsl::minimum_temperature,
+                    temperature_breach_config_dsl::maximum_temperature,
+                ),
+            ))
+            .filter(temperature_log_dsl::temperature_breach_id.is_null())
+            .order(temperature_log_dsl::datetime.asc())
+            .into_boxed();
+
+        apply_equal_filter!(query, filter.store_id, temperature_log_dsl::store_id);
+        apply_date_time_filter!(query, filter.datetime, temperature_log_dsl::datetime);
+
+        query = query.filter(temperature_breach_config_dsl::is_active.eq(true));
+
+        apply_equal_filter!(
+            query,
+            Some(TemperatureBreachRowType::Excursion.equal_to()),
+            temperature_breach_config_dsl::type_
+        );
+
+        // Debug diesel query
+        // println!("{}", diesel::debug_query::<DBType, _>(&query).to_string());
+
+        let mut excursion_data: Vec<TemperatureExcursionRow> = Vec::new();
+        let log_data = query
+            .load::<QueryResult>(&self.connection.connection)?
+            .into_iter()
+            .map(TemperatureRow::from)
+            .collect::<Vec<TemperatureRow>>();
+
+        for row in log_data.iter() {
+            if row.is_excursion == true {
```

Comment:
I had this as a single sql query, with the thought of running the logic in SQL and returning a small result set.
That ended up being fine in pg, but slow in sqlite, and I also had trouble debugging after implementing in diesel, so I scrapped that and moved to running the logic in rust, working with a Vec of temperature logs. 

The date filter is intended to reduce the dataset size - will run more testing once I hook this up and see how it performs in pg / sqlite with the data I have..